drivers: nrf_rtc_timer: force init of interrupt and event enabling

giansta opened the pull request:
Make sure that the interrupt and event routing enable bits have the reset values in case the Zephyr firmware image is chain-loaded.

When started by MCUboot, these registers values aren't the reset values, but then Zephyr initializes them in same way. Yet, when the Zephyr firmware image is chain-loaded by nRF5 bootloader, they may have values that cause the scheduler to behave in unexpected ways.

nordicjm commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -667,6 +674,15 @@ static int sys_clock_driver_init(const struct device *dev)
 			CLOCK_CONTROL_NRF_LF_START_AVAILABLE :
 			CLOCK_CONTROL_NRF_LF_START_STABLE);
 
+#if defined(CONFIG_INIT_ARCH_HW_AT_BOOT)
```
This is not valid, this Kconfig is for setting the default architecture peripherals and registers to known values, architecture in this case is ARMv7E-M - the ARM architecture has no knowledge of the nordic rtc peripheral because that is not part of the ARM architecture, and it would be wrong to start adding SoC-specific handling to this to address an issue with 1 bootloader.

giansta replied:
Thanks! So, would it be right is I just remove the #if directive and change the commit message accordingly?

anangl commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -40,6 +40,13 @@ BUILD_ASSERT(CHAN_COUNT <= RTC_CH_COUNT, "Not enough compare channels");
 #define ANCHOR_RANGE_END (7 * COUNTER_SPAN / 8)
 #define TARGET_TIME_INVALID (UINT64_MAX)
 
+#define NRF_RTC_INT_ALL_MASK (RTC_INTENSET_TICK_Msk |		\
```
The `NRF_RTC_` prefix should be reserved for stuff defined in `<hal/nrf_rtc.h>`. Please use some other name.

giansta replied:
Thanks, sorry for the mess. Indeed there was no need of a new define and I used now only the defines in the interface nrf_rtc.h.

anangl commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -40,6 +40,13 @@ BUILD_ASSERT(CHAN_COUNT <= RTC_CH_COUNT, "Not enough compare channels");
 #define ANCHOR_RANGE_END (7 * COUNTER_SPAN / 8)
 #define TARGET_TIME_INVALID (UINT64_MAX)
 
+#define NRF_RTC_INT_ALL_MASK (RTC_INTENSET_TICK_Msk |		\
+			      NRF_RTC_INT_OVERFLOW_MASK |	\
```
Please don't mix RTC HAL definitions with those from MDK. Use `NRF_RTC_*` symbols consistently.

giansta replied:
same as above

nordicjm commented on drivers/timer/nrf_rtc_timer.c:
```diff
@@ -40,6 +40,13 @@ BUILD_ASSERT(CHAN_COUNT <= RTC_CH_COUNT, "Not enough compare channels");
 #define TARGET_TIME_INVALID (UINT64_MAX)
 
+#define NRF_RTC_INT_ALL_MASK (RTC_INTENSET_TICK_Msk |		\
+			      NRF_RTC_INT_OVERFLOW_MASK |	\
```
Use the same type of named variables here, replace the _MASK one with a _Msk one

giansta replied:
same as above

giansta committed:
drivers: nrf_rtc_timer: force init of interrupt and event enabling

Make sure that the interrupt and event routing enable bits have the
reset values in case the Zephyr firmware image is chain-loaded.
When started by MCUboot, these registers values aren't the reset values,
but then Zephyr initializes them in same way. Yet, when the Zephyr
firmware image is chain-loaded by nRF5 bootloader, they may have values
that cause the scheduler to behave in unexpected ways.

Signed-off-by: Giancarlo Stasi <giancarlo.stasi.co@gmail.com>
